sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryServicesImpl.java

```diff
@@ -1001,4 +1007,141 @@ public void close() {
       client.close();
     }
   }
+
+  private static class BoundedExecutorService implements ExecutorService {
+    private final ExecutorService executor;
+    private final Semaphore semaphore;
+    private final int parallelism;
+
+    BoundedExecutorService(ExecutorService executor, int parallelism) {
+      this.executor = executor;
+      this.parallelism = parallelism;
+      this.semaphore = new Semaphore(parallelism);
+    }
+
+    @Override
+    public void shutdown() {
+      executor.shutdown();
+    }
+
+    @Override
+    public List<Runnable> shutdownNow() {
+      List<Runnable> runnables = executor.shutdownNow();
+      // try to release permits as many as possible before returning semaphored runnables.
+      synchronized (this) {
+        if (semaphore.availablePermits() <= parallelism) {
+          semaphore.release(Integer.MAX_VALUE - parallelism);
```

Comment:
explain this logic? Why Integer.MAX_VALUE?

Reply:
I wanted to remove the restriction from the semaphore by setting the number of allowed permits to its maximum value (which is Integer.MAX_VALUE). Looks like it's impossible to retrieve the original Callable here so returning maximally allowed semaphored callable instead.

Reply:
I just want to ensure that this is legal. Releasing more permits than acquired seems a bit weird, and I'm worried that we will hit some edge case. Have you verified that this is ok?

Reply:
I think we don't have to pair acquire() and release(). Excerpted from release() API doc:

> There is no requirement that a thread that releases a permit must have acquired that permit by calling acquire().
> https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/Semaphore.html#release--

The possible edge case would be that if we put the total number of permits more than Integer.MAX_VALUE by calling release() then it throws an exception. By checking availablePermits() before release() in synchronized section we can avoid those cases.

Other option here is we can just return semaphored callables as is and document it clearly in a comment. I believe that this `BoundedExecutorService` class will hardly be reused anyway.